docs/csharp/ref-locals-and-returns.md

```diff
@@ -0,0 +1,205 @@
+
+## `ref readonly` returns
+
+You may also want to return a value type by reference, but disallow the
+caller from modifying that value. Use the `ref readonly` modifier to
+express that design intent. It notifies readers
+that you are returning a reference to existing data, but not allowing
+modification. 
+
+The compiler enforces that the caller cannot modify the reference. Attempts
+to assign to the value directly generate a compile-time error. Calling member
+methods creates a defensive copy. 
+
+Creating a copy of a ref readonly return is easy: Just assign it to a variable
+not declared with the `ref readonly` modifier. The compiler generates code
+to copy the object as part of the assignment. 
+
+It's likely that the library using `Point3D` would often use the origin
+throughout the code. Every instance creates a new object on the stack. It may
+be advantageous to create a constant and return it by reference. But, if you
+are returning a reference to internal storage, you may want to enforce that
+the caller cannot modify the referenced storage. The following code
+defines a read-only property that returns a `readonly ref` to a `Point3D`
+that specifies the origin.
+
+[!code-csharp[OriginReference](../../samples/csharp/reference-semantics/Point3D.cs#OriginReference "Creating a readonly Origin reference")]
+
+When you assign a variable to a `ref return`, you can specify either a `ref readonly`
+variable, or a by-value copy of the readonly reference:
+
+[!code-csharp[AssignRefReadonly](../../samples/csharp/reference-semantics/Program.cs#AssignRefReadonly "Assigning a ref readonly")]
+
+The first assignment in the preceding code makes a copy of the `Origin` constant, and assigns
+that copy. The second assigns a reference. Notice that the `readonly` modifier
+must be part of the declaration of the variable. The reference to which it refers
+cannot be modified. Attempts to do so results in a compile-time error.
+
+## `readonly struct` type
+
+Applying `ref readonly` to high-traffic uses of a struct may be sufficient.
+Other times, you may want to create an immutable struct. Then, you can
+always pass by readonly reference. That practice 
+removes the defensive copies
+that take place when you access methods of a struct used as an `in` parameter.
+
+You can do that by creating a `readonly struct` type. You can add the `readonly`
+modifier to a struct declaration. The language enforces that all members of
+the struct are `readonly`: the `struct` must be immutable.
+
+There are other optimizations for a `readonly struct`. You can
+use the `in` modifier at every location where a `readonly struct` is an
+argument. In addition, you can return a `readonly struct` as a `ref return` when
+you are returning an object whose lifetime extends beyond the scope of the method
+returning the object.
+
+Finally, the compiler generates more efficient code when you call members of a
+`readonly struct`: The `this` reference is always an `in` parameter, instead
```

Comment:
Hard to parse. Perhaps "The `this` reference, instead of a copy of the receiver, is always an `in` parameter passed by value to the member method.

Reply:
Also, it should be "passed by reference" not "passed by value".